Docs for dbt-databricks v1.10.0 (#7120)
fishtown-analytics/docs.getdbt.com: website/docs/reference/global-configs/databricks-changes.md

Setting this flag to `True` will use `information_schema` rather than `describe extended` to get column metadata for Unity Catalog tables. This setting helps you avoid issues where `describe extended` truncates information when the type is a complex struct. However, this setting is not yet the default behavior, as there are performance impacts due to a Databricks metadata limitation because of the need to run `REPAIR TABLE {{relation}} SYNC METADATA` before querying to ensure the `information_schema` is complete. 
Please note that there is no equivalent option for views at this time which means dbt will still need to use `describe extended` for views.

This flag will become the default behavior when this additional query is no longer needed.

## Use user's folder for Python model notebooks

The `use_user_folder_for_python` flag is `False` by default and results in writing uploaded python model notebooks to `/Shared/dbt_python_models/{{schema}}/`. Setting this flag to `True` will write notebooks to `/Users/{{current user}}/{{catalog}}/{{schema}}/` Writing to the `Shared` folder is deprecated by Databricks as it does not align with governance best practices.

> We plan to switch the default of this flag to `True` in v1.11.0.

Reviewer: ```suggestion
When set to `True`, `dbt-databricks ` uses the updated logic for all model types (views, tables, incremental, seeds) and optionally enables additional config options for more fine-tuned control.

We plan to switch the default of this flag to `True` in v1.11.0.
```

## Use restructured materializations

The `use_materialization_v2` flag is `False` by default and guards significant rewrites of the core materializations in `dbt-databricks` while they are still in an experimental stage.

> We plan to switch the default of this flag to `True` in 1.11.0, depending on user feedback.

Reviewer: i think it might be helpful here to explicitly state what happens when true. and also introduce how it affects other things like views and seeds
```suggestion
When set to `True`, `dbt-databricks ` uses the updated logic for all model types (views, tables, incremental, seeds) and optionally enables additional config options for more fine-tuned control.

We plan to switch the default of this flag to `True` in 1.11.0, depending on user feedback.
```

> 

Reviewer: i think it'd also be helpful to state the new optional model configs so it transitions into the next sections. so for example:

> When setting the `use_materialization_v2` flag to `True`, it unlocks two optional model configurations for more control over how to handle existing relations:
- `view_update_via_alter` &mdash; When enabled, this config attempts to update the view in place using `alter view`, instead of using `create or replace` to replace it. This allows continuity of history for the view, keeps the metadata, and helps with Unity Catalog compatibility. 
- `use_safer_relation_operations` &mdash; When enabled (and if `view_update_via_alter` isn't set), this config makes dbt update models more safe by creating the new relation in a staging location, swapping it with the existing view, deleting the old relation afterward, and avoids `create or replace...` operations.

You don’t need to use these configs as they’re optional and only relevant when you enable the `use_materialization_v2` flag.

Reviewer 2: @benc-db  could we make this change?

Author: Oh, I think I misinterpreted your feedback the first time, let me take another look

Author: Please take another look

> We currently plan to switch the default of this flag to `True` in 1.11.0, depending on user feedback.

Reviewer: ```suggestion
We plan to switch the default of this flag to `True` in 1.11.0, depending on user feedback.
```

### Changes to the Seed materialization

The seeds materialization should have the smallest difference between the old and new materialization, as the primary difference is just removing calls to methods that are not supported by Databricks, such as transaction operations.

### Changes to the View materialization

In conjunction with the flag above, there are two model configuration options that can customize how we handle the view materialization when we detect an existing relation at the target location.

#### `view_update_via_alter`

For view models where we set `view_update_via_alter: True`, when we detect an existing view, rather than using `create or replace...` to replace the existing view, we gather metadata to determine if the existing view matches the dbt project.
When it does, no further action is taken; otherwise, `alter view...` is used to alter the view in place to match your dbt project.
This allows continuity of history for the view, which is lost with `create or replace...`, which can fix user-reported issues with Unity Catalog features.

> 

Reviewer: i think for both model configs, it'd be great to have an example of the behavior flag enabled, as well as the model configs enabled so users know how this config looks?

Author: Given the other feedback and responses, where would you place the examples?

Reviewer: that's a good question @benc-db !  i recommend adding a short “Example” block at the end of each subsection after the model configuration explanation. i think this gives users immediate, relevant context after they’ve learned about the config’s purpose.

Reviewer: <img width="767" alt="Screenshot 2025-04-03 at 15 12 28" src="https://github.com/user-attachments/assets/5782ea4b-988d-490e-be53-97eb69221d1e" />
<img width="789" alt="Screenshot 2025-04-03 at 15 12 35" src="https://github.com/user-attachments/assets/dcdb31b3-b034-4294-b5d3-7c17312984a8" />

:::caution There is currently no support for altering the comment on a view via Databricks SQL.

As such, we must replace the view whenever you change its description

:::

#### `use_safer_relation_operations`

The other new model configuration introduced for views is `use_safer_relation_operations`.

> When this configuration is set to `True`, if `view_update_via_alter` is not set, we will use a multi-step, renaming approach to replacing existing relations.

Reviewer: what happens if you set `view_update_via_alter` as well?

Author: view_update_via_alter supersedes use_safer_relation_operations if there is a pre-existing view; if there is a pre-existing relation and its not a view, or there is not a pre-existing relation, use_safer_relation_operations can apply (create the view in a staging location, swap out using renames if able).

Rather than using a `create or replace...` statement to replace an existing view, we create the new view at a staging location, then swap locations with the existing view, before finally deleting the existing view.

:::caution This configuration option may increase storage costs.

While this approach is equivalent to the default dbt view materialization, it will create additional UC objects, as compared to alternatives.

:::

While this setting is more relevant for table and incremental materializations, there are benefits for use with views.
For example, when replacing a materialized view with a traditional view, we have historically dropped the materialized view prior to creating the new view.
Although materialized views do not currently support renaming, setting this config ensures that the materialized view is not dropped until we have successfully created the new view in the staging location; as such, there is no longer a scenario where the existing relation is dropped and the new relation fails to be created.

> 

Reviewer: i think for both model configs, it'd be great to have an example of the behavior flag enabled, as well as the model configs enabled so users know how this config looks?

| ----------------------------- | ----------------------- | -------------------------- |
| `use_info_schema_for_columns` | 1.9.0                   | TBD                        |
| `use_user_folder_for_python`  | 1.9.0                   | TBD                        |
| `use_materialization_v2`      | 1.10.0                  | TBD                        |

## Use information schema for columns

The `use_info_schema_for_columns` flag is `False` by default.

> Setting this flag to `True` will use `information_schema` rather than `describe extended` to get column metadata for Unity Catalog tables. This setting helps you avoid issues where `describe extended` truncates information when the type is a complex struct. However, this setting is not yet the default behavior, as there are performance impacts due to a Databricks metadata limitation because of the need to run `REPAIR TABLE {{relation}} SYNC METADATA` before querying to ensure the `information_schema` is complete. Furthermore, there is no equivalent option for views at this time.

Reviewer 2: ```suggestion
Setting this flag to `True` will use `information_schema` rather than `describe extended` to get column metadata for Unity Catalog tables. This setting helps you avoid issues where `describe extended` truncates information when the type is a complex struct. However, this setting is not yet the default behavior, as there are performance impacts due to a Databricks metadata limitation because of the need to run `REPAIR TABLE {{relation}} SYNC METADATA` before querying to ensure the `information_schema` is complete. 
Please note that there is no equivalent option for views at this time which means dbt will still need to use `describe extended` for views.
```

is this what you mean?

Author: Yes.  There is no 'REPAIR VIEWS' to sync metadata.

Related to this, I learned something very valuable this week and would like your opinion on where/how to handle it in docs.  A lot of times when users have complex column types, its because they are processing json data, say from a production system structured log.  They are using from_json and getting a complex struct/array type out.  This week I recommended a user tryout replacing with parse_json, which gives the variant type back, and voila, no more complex type.  So basically, I'm wondering where/how we should communicate that to users?  Then the problem this flag is trying to solve is avoided.

## Use user's folder for Python model notebooks

The `use_user_folder_for_python` flag is `False` by default and results in writing uploaded python model notebooks to `/Shared/dbt_python_models/{{schema}}/`. Setting this flag to `True` will write notebooks to `/Users/{{current user}}/{{catalog}}/{{schema}}/` Writing to the `Shared` folder is deprecated by Databricks as it does not align with governance best practices.

We plan to switch the default of this flag to `True` in v1.11.0.

## Use restructured materializations

The `use_materialization_v2` flag is `False` by default and guards significant rewrites of the core materializations in `dbt-databricks` while they are still in an experimental stage.

We currently plan to switch the default of this flag to `True` in 1.11.0, depending on user feedback.

### Changes to the Seed materialization

The seeds materialization should have the smallest difference between the old and new materialization, as the primary difference is just removing calls to methods that are not supported by Databricks, such as transaction operations.

### Changes to the View materialization

> In conjunction with the flag above, there are two model configuration options that can customize how we handle the view materialization when we detect an existing relation at the target location.

Reviewer 2: does this mean they have to use the flag and also these two model configs below?

Reviewer 2: the nesting of flags isn't the nicest experience. 🤔

Author: They don't need to use either of them.  These are new model config options that are opened up if they opt into the new behavior.  Essentially I have one flag that gates experimental materialization rewrites.  When the flag is false, they get materializations that match 1.9.  In the new materializations, there are new options for how to materialize the models.  I'm open to revising this, was just looking to make all of the new materialization behavior opt-in initially, since they are such a fundamental change.

Reviewer 2: I see - I was confused about how things were nested in the doc. No need to change the behavior

#### `view_update_via_alter`

> For view models where we set `view_update_via_alter: True`, when we detect an existing view, rather than using `create or replace...` to replace the existing view, we gather metadata to determine if the existing view matches the dbt project.

Reviewer 2: This will be a longstanding flag right? This sounds more like a flag that is more like a global config not a behavior flag?

Author: This is a model config that is gated by the behavior flag above.  This model config requires use_materialization_v2 to be True in order to do anything.  Happy to restructure.  Will answer above comment too.

Author: I guess the question is really, do we think users would want to choose different view behavior on a per-model basis?  As currently implemented, they have that flexibility, but it may not be necessary; having said that, I'm not really sure I understand how to implement global flags that are not behavior flags.

Reviewer 2: I can see arguments for both but it would be easier to understand projects with just one behavior. 
I'll just resolve this for now

Another benefit of this approach is that several other column features, such as column-level masks, are also incompatible with inserting data as part of create.
While these features are not a part of the 1.10.0 release, they can now be supported in future releases.

#### Constraints

For several feature releases now, dbt-databricks supported both the dbt implementation of constraints and an alternative called `persist_constraints`, that we had implemented prior.
With the `use_materialization_v2` flag, we begin the deprecation of this alternative implementation, and expand our support for the core dbt implementation.
We now support the expression field for primary and foreign keys as a way to supply additional Databricks options, such as using [`RELY` to tell the Databricks optimizer that it may exploit the constraint to rewrite queries](https://docs.databricks.com/aws/en/sql/language-manual/sql-ref-syntax-ddl-create-table-constraint).

> Another change to how constraints work in practice is due to separating create and insert.

Reviewer 2: ```suggestion
Another change to how constraints work in practice is due to separating out `create` and `insert`.
```

| ----------------------------- | ----------------------- | -------------------------- |

> | `use_info_schema_for_columns` | 1.9.0                   | TBD                        |

Reviewer: ```suggestion
| [`use_info_schema_for_columns`](#use-information-schema-for-columns) | 1.9.0                   | TBD                        |
```

> | `use_user_folder_for_python`  | 1.9.0                   | TBD                        |

Reviewer: ```suggestion
| [`use_user_folder_for_python`](#use-users-folder-for-python-model-notebooks)  | 1.9.0                   | TBD                        |
```

> | `use_materialization_v2`      | 1.10.0                  | TBD                        |

Reviewer: ```suggestion
| [`use_materialization_v2`](#use-restructured-materializations)      | 1.10.0                  | TBD                        |
```

## Use information schema for columns

The `use_info_schema_for_columns` flag is `False` by default.

Setting this flag to `True` will use `information_schema` rather than `describe extended` to get column metadata for Unity Catalog tables. This setting helps you avoid issues where `describe extended` truncates information when the type is a complex struct. However, this setting is not yet the default behavior, as there are performance impacts due to a Databricks metadata limitation because of the need to run `REPAIR TABLE {{relation}} SYNC METADATA` before querying to ensure the `information_schema` is complete. 
Please note that there is no equivalent option for views at this time which means dbt will still need to use `describe extended` for views.

> This flag will become the default behavior when this additional query is no longer needed.

Reviewer: which additional query do we mean here?

Author: The 'REPAIR TABLE' command.  And honestly, it really comes down to how Databricks's metadata strategy evolves.  We could end up pivoting depending on whether or not Databricks makes needed improvements to information schema.
